### PR TITLE
THNN functional module conversions batch 4

### PR DESCRIPTION
**Pull request by andreaskoepf**
- [x] MarginCriterion
- [x] MSECriterion
- [x] MultiLabelMarginCriterion
- [x] MultiMarginCriterion
- [x] PReLU
- [x] RReLU
- [x] Sigmoid
- [x] SmoothL1Criterion
- [x] SoftMax
- [x] SoftPlus
- [x] SoftShrink
- [x] Sqrt
- [x] Square
- [x] Tanh
- [x] Threshold
- [x] Cross-check with old module implementation via [xtest.lua](https://github.com/andreaskoepf/nn/blob/nn2/xtest.lua#L343)

### CODE REVIEW

File: MultiMarginCriterion.lua [hunk 1]
```diff
@@ -9,10 +9,36 @@ end
 
 function MultiMarginCriterion:updateOutput(input, target)
    -- backward compatibility
+   local _target = target
+   if not torch.isTensor(_target) then
+     _target = input.new(1)
```
Comment (fmassa) [inline MultiMarginCriterion.lua]: if `target` is a number, this allocates memory at each iteration (although it's only a 1-element tensor...)

File: MultiMarginCriterion.lua [hunk 2]
```diff
@@ -9,10 +9,36 @@ end
...
+   )
+   self.output = self.output_tensor[1]
+   return self.output
 end
 
 function MultiMarginCriterion:updateGradInput(input, target)
-   return input.nn.MultiMarginCriterion_updateGradInput(self, input, target)
+   local _target = target
+   if not torch.isTensor(_target) then
+     _target = input.new(1)
```
Comment (fmassa) [inline MultiMarginCriterion.lua]: same comment here

File: MultiMarginCriterion.lua (hunk 1 above)
Comment (andreaskoepf) [inline MultiMarginCriterion.lua]: I moved the 'tensorization' from the old C implementation to lua to have a single signature. I will add the classical lua 'or' conditional assignment trick.

**Comment by andreaskoepf** [comment]
Note for reviewers: Unfortunately the total changes listed under `Files changed` do not show all line-changes because some files have probably been treated as 'rewritten'. To see the file changes line by line please click through the individual commits.

**Comment by soumith** [comment]
cc: @hughperkins

**Comment by hughperkins** [comment]
This looks hard.  Will need a few weeks to look into this.

**Comment by soumith** [comment]
@andreaskoepf now this has conflicts.

**Comment by andreaskoepf** [comment]
I rebased this PR, also updated https://github.com/torch/cunn/pull/198, should be ready for merge

**Comment by hughperkins** [comment]
Ok. clnn is ready for merge, for this PR, https://github.com/hughperkins/clnn/tree/thnn4

**Comment by hughperkins** [comment]
Thats odd, the underlying `nn` no longer builds for me:

```
Scanning dependencies of target THNN
[100%] Building C object lib/THNN/CMakeFiles/THNN.dir/init.c.o
In file included from /home/ubuntu/torch/install/include/TH/TH.h:4:0,
                 from /home/ubuntu/git/nn/lib/THNN/init.c:1:
/home/ubuntu/git/nn/lib/THNN/generic/MultiLabelMarginCriterion.c: In function ‘THNN_FloatMultiLabelMarginCriterion_updateOutput’:
/home/ubuntu/torch/install/include/TH/THTensor.h:8:39: error: too many arguments to function ‘THFloatTensor_minall’
 #define THTensor_(NAME)   TH_CONCAT_4(TH,Real,Tensor_,NAME)
                                       ^
/home/ubuntu/torch/install/include/TH/THGeneral.h:60:57: note: in definition of macro ‘THArgCheck’
 #define THArgCheck(...) _THArgCheck(__FILE__, __LINE__, __VA_ARGS__)
                                                         ^
/home/ubuntu/torch/install/include/TH/THGeneral.h:90:30: note: in expansion of macro ‘TH_CONCAT_4_EXPAND’
 #define TH_CONCAT_4(x,y,z,w) TH_CONCAT_4_EXPAND(x,y,z,w)
```

```
nn$ git status
On branch torch-master
Your branch is up-to-date with 'origin/master'.
Untracked files:
  (use "git add <file>..." to include in what will be committed)

	generic/Linear.cpp
	share

nothing added to commit but untracked files present (use "git add" to track)
ubuntu:~/git/nn$ git log -n 5 --oneline
b2af7ea fix for torch7 minall/maxall changes
89017f4 Merge pull request #578 from andreaskoepf/THNN_koepf_4
efee63e Merge pull request #600 from ivpopov/patch-2
dcb4f3d Test nn.CMul for non-contiguous tensors
83a17b2 Add THNN conversion of {oftShrink, Sqrt, Square, Tanh, Threshold}
```

**Comment by hughperkins** [comment]
If I `git checkout 89017f4`, then it works, so the max/min commit seems at fault.

**Comment by soumith** [comment]
fixed the max/min issue in nn trunk

**Comment by hughperkins** [comment]
I did a `git pull` but no change? 
```
ubuntu:~/git/nn$ git remote -v
andreas	https://github.com/andreaskoepf/nn.git (fetch)
andreas	https://github.com/andreaskoepf/nn.git (push)
fmassa	git@github.com:fmassa/nn.git (fetch)
fmassa	git@github.com:fmassa/nn.git (push)
hugh	git@github.com:hughperkins/nn.git (fetch)
hugh	git@github.com:hughperkins/nn.git (push)
origin	git@github.com:torch/nn.git (fetch)
origin	git@github.com:torch/nn.git (push)
ubuntu:~/git/nn$ git fetch origin
ubuntu:~/git/nn$ git checkout origin/master
Note: checking out 'origin/master'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at b2af7ea... fix for torch7 minall/maxall changes
ubuntu:~/git/nn$ git log -n 1 --oneline
b2af7ea fix for torch7 minall/maxall changes
```

**Comment by soumith** [comment]
hmmm, looks like you do get https://github.com/torch/nn/commit/b2af7eaddfc1de72661b6861115d3fdb97403bf3 , and that should fix nn builds and tests.

**Comment by hughperkins** [comment]
Hmmm.  Maybe it's  a gcc-4.9 thing.

**Comment by hughperkins** [comment]
oh... it's using include files from ~/torch.  Maybe I will delete those, and retry

**Comment by hughperkins** [comment]
Ah, my torch7 was too old :-)

**Comment by andreaskoepf** [comment]
@soumith thnx for fixin the THFloatTensor_minall issue. It's a bit strange that the continous integration build + unit test did not catch this. Was there a recent change to THFloatTensor_minall()?

**Comment by soumith** [comment]
@andreaskoepf yea was a change + revert today morning in a narrow window.

**Pull request merged**